lisa/util/shell.py

```diff
@@ -44,25 +45,103 @@ def __init__(
             raise LisaException("username must be set")
 
 
+class WindowsShellType(object):
+    """
+    Windows command generator
+    Support get pid, set envs, and cwd
+    Doesn't support kill, it needs overwrite spur.SshShell
+    """
+
+    supports_which = False
+
+    def generate_run_command(
+        self,
+        command_args: List[str],
+        store_pid: bool = False,
+        cwd: Optional[str] = None,
+        update_env: Optional[Dict[str, str]] = None,
+        new_process_group: bool = False,
+    ) -> str:
+        commands = []
+
+        if store_pid:
+            commands.append(
+                'powershell "(gwmi win32_process|? processid -eq $pid).parentprocessid"'
+                " &&"
+            )
+
+        if cwd is not None:
+            commands.append(f"cd {cwd} 2>&1 && echo spur-cd: 0 ")
+            commands.append("|| echo spur-cd: 1 && exit 1 &")
+
+        if update_env:
+            update_env_commands = [
+                "set {0}={1}".format(key, value) for key, value in update_env.items()
+            ]
+            commands += f"{'; '.join(update_env_commands)}; "
+
+        if cwd is not None:
+            commands.append(f"pushd {cwd} & ")
+            commands.append(" ".join(command_args))
+            commands.append(" & popd")
+        else:
+            commands.append(" ".join(command_args))
+        return " ".join(commands)
+
+
 class SshShell:
     def __init__(self, connection_info: ConnectionInfo) -> None:
         self.is_remote = True
         self._is_initialized = False
         self._connection_info = connection_info
-        self._inner_shell: Optional[spurplus.SshShell] = None
+        self._inner_shell: Optional[spur.SshShell] = None
 
         paramiko_logger = getLogger("paramiko")
         paramiko_logger.setLevel(logging.WARN)
 
     def initialize(self) -> None:
-        self._inner_shell = spurplus.connect_with_retries(
-            self._connection_info.address,
-            port=self._connection_info.port,
-            username=self._connection_info.username,
-            password=self._connection_info.password,
-            private_key_file=self._connection_info.private_key_file,
-            missing_host_key=spur.ssh.MissingHostKey.accept,
+        # spur try a linux command and return error on Windows.
+        # So try with paramiko firstly.
+        paramiko_client = paramiko.SSHClient()
+        paramiko_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
```

Comment:
Even though LISA is only to be used in controlled instances, it might not be a good idea to blindly accept any host keys.  For the case where tests are executed in an Azure VM, is there a way to get the host key with an API call prior to connecting and rejecting anything that doesn't match it?

Reply:
Not sure if azure have such one, but lisa supports also existing servers, hyperv and others. To get full automated, and reduce bar to run cases, so accept all servers.